File: SquaresDemo/Geometry/Point2d.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace SquaresDemo.Geometry
{
    public class Point2d
    {
        public static Point2d Zero { get; } = new Point2d(0, 0);

        public int X { get; }
        public int Y { get; }

        public Point2d(int x, int y)
        {
            X = x;
            Y = y;
        }

        public Point2d Add(Point2d other) => new Point2d(X + other.X, Y + other.Y);

        public Point2d Subtract(Point2d other) => new Point2d(X - other.X, Y - other.Y);

        public int DotProduct(Point2d other) => (X * other.X + Y * other.Y);

        public override string ToString() => $"({X},{Y})";
    }
}

File: SquaresDemo/scripts/point2d.js
function Point2D(x, y, id) {
    this.x = x;
    this.y = y;
    this.id = id;
}

Point2D.MinX = -5000;
Point2D.MaxX = 5000;

Point2D.MinY = -5000;
Point2D.MaxY = 5000;

Point2D.Validate = function (x, y) {
    return Point2D.MinX <= x && x <= Point2D.MaxX && Point2D.MinY <= y && y <= Point2D.MaxY;
}

Point2D.prototype.Add = function(other){
    return new Point2D(this.x + other.x, this.y + other.y);
}

Point2D.prototype.Subtract = function(other) {
    return new Point2D(this.x - other.x, this.y - other.y);
}

Point2D.prototype.DotProduct = function(other) {
    return this.x * other.x + this.y * other.y;
}

export default Point2D;